File: src/hilbert/nodes/planner.py
"""Planner node for Hilbert."""

from hilbert.llm import get_client, get_planner_prompt, get_dimensions_fallback
from hilbert.llm.utils import parse_json_object
from hilbert.state.research import ResearchState


async def planner_node(state: ResearchState) -> dict:
    """Decompose query into sub-questions and four research dimensions."""
    query = state["query"]
    max_rounds = state["max_rounds"]
    n = max(max_rounds * 2, 4)

    callback = state.get("progress_callback")
    if callback:
        callback("planner", {"status": "planning", "query": query})

    system_prompt, user_prompt = get_planner_prompt(query, n=n)

    sub_questions = [query]
    dimensions = get_dimensions_fallback(query)

    try:
        client = get_client()
        content = await client.complete_text(
            prompt=user_prompt,
            system_prompt=system_prompt,
        )

        parsed = parse_json_object(content)
        if parsed:
            qs = parsed.get("sub_questions", [])
            dims = parsed.get("dimensions", [])
            if qs:
                sub_questions = [str(q) for q in qs]
            if dims and len(dims) == 4:
                dimensions = dims

    except Exception:
        pass  # keep fallbacks

    return {
        "sub_questions": sub_questions,
        "research_dimensions": dimensions,
        "round": 1,
        "status": "searching",
    }


def create_planner_node():
    """Create planner node function."""
    return planner_node

File: cli/src/core/nodes/planner.ts
import { getClient } from "../llm/client";
import { getPlannerPrompt, getDimensionsFallback, type ResearchDimension } from "../llm/prompts";
import { parseJsonObject } from "../llm/utils";
import type { ResearchState } from "../state";

const STRATEGY_AUGMENTS: Record<string, string> = {
  applied: "experiment evaluation benchmark",
  survey: "survey review overview",
};

const ARXIV_DATE_FILTER: Record<string, string> = {
  recent: " AND submittedDate:[20230101 TO *]",
};

export async function plannerNode(state: ResearchState): Promise<Partial<ResearchState>> {
  const { query, maxRounds } = state;
  const n = Math.max(maxRounds * 2, 4);

  const callback = state.progressCallback;
  if (callback) {
    callback("planner", { status: "planning", query });
  }

  const [systemPrompt, userPrompt] = getPlannerPrompt(query, n);

  let subQuestions = [query];
  let dimensions = getDimensionsFallback(query);

  try {
    const client = getClient();
    const content = await client.completeText(userPrompt, systemPrompt);
    const parsed = parseJsonObject(content);

    if (parsed) {
      const qs = parsed.sub_questions as string[] | undefined;
      const dims = parsed.dimensions as Record<string, unknown>[] | undefined;
      if (qs && qs.length > 0) {
        subQuestions = qs.map(String);
      }
      if (dims && dims.length === 4) {
        dimensions = dims as unknown as ResearchDimension[];
      }
    }
  } catch {
    // keep fallbacks
  }

  return {
    subQuestions,
    researchDimensions: dimensions,
    round: 1,
    status: "searching" as const,
  };
}

export function createPlannerNode() {
  return plannerNode;
}